### PR TITLE
Fusion 5 dev

### PR DESCRIPTION
**Pull request by sudosoup**
foo

### CODE REVIEW

File: projects/buttons/Makefile [hunk 1]
```diff
@@ -19,7 +19,7 @@ push:
 
 upload:
 	git pull
-	gcloud app deploy app.yaml --project labs-209320
+	gcloud app deploy app.yaml --project labs-3-datastore-dep
```
Comment (MarcusSorealheis) [inline projects/buttons/Makefile]: needs to change this val for prod.

File: projects/buttons/fastener/main.py [hunk 2]
```diff
@@ -12,7 +12,7 @@
 import re
 
 # generators
-def id_generator(size=4, chars=string.ascii_lowercase + string.digits):return ''.join(random.choice(chars) for _ in range(size))
+def id_generator(size=4, chars=string.ascii_lowercase + "labs"):return ''.join(random.choice(chars) for _ in range(size))
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/main.py]: can we up the size here to 10

File: projects/buttons/fastener/main.py [hunk 3]
```diff
@@ -25,12 +25,12 @@ def password_generator(size=12, chars=string.ascii_lowercase + string.digits):re
...
         if 'token' in item:
             key,token = item.split('-')
-            
+
 # google creds
 credentials = compute_engine.Credentials()
 compute = discovery.build('compute', 'v1', credentials=credentials)
 compute_beta = discovery.build('compute', 'beta', credentials=credentials)
-project = 'labs-209320'
+project = 'labs-3-datastore-dep'
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/main.py]: change to appropriate project

File: projects/buttons/fastener/main.py [hunk 4]
```diff
@@ -439,7 +456,7 @@ def create(stream_slug='lou'):
         "items": [
         {
             "key": "startup-script-url",
-            "value": "https://raw.githubusercontent.com/lucidworks/streams/master/projects/buttons/fastener/scripts/start-button.sh"
+            "value": "https://raw.githubusercontent.com/sudosoup/streams/fusion-5-dev/projects/buttons/fastener/scripts/start-button.sh"
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/main.py]: This needs to change to point the correct branch and repo. Make this change absolutely last.

File: projects/buttons/fastener/scripts/start-button.sh [hunk 5]
```diff
@@ -34,7 +34,7 @@ fi
 
 IP=$(wget -qO- http://ipecho.net/plain)
 
-cd /; git clone https://github.com/lucidworks/streams
+cd /; git clone https://github.com/sudosoup/streams
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/scripts/start-button.sh]: this one as well. Should clone from upstream.

File: projects/buttons/fastener/scripts/start-button.sh [hunk 6]
```diff
@@ -20,7 +20,7 @@ echo "IID: $IID"
 # etc....
 # }
 
-STREAM_JSON=`curl https://streams.lucidworks.com/api/stream/$SID`
+STREAM_JSON=`curl https://labs-3-datastore-dep.appspot.com/api/stream/$SID`
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/scripts/start-button.sh]: remember to point at streams.

File: projects/buttons/fastener/scripts/start-button.sh [hunk 7]
```diff
@@ -87,7 +87,7 @@ if [ ! -d "/fusion" ]; then
 # # if fusion not installed
 # #############################
 
-gsutil cp gs://buttons-streams/fusion-${FUSION_VERSION}.tar.gz .
+gsutil cp gs://buttons-dev/fusion-${FUSION_VERSION}.tar.gz .
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/scripts/start-button.sh]: Does the stream exist in the main project.

File: projects/buttons/fastener/scripts/start-button.sh [hunk 8]
```diff
@@ -133,7 +133,16 @@ else
...
 fi
 
-# TODO: check for existence (and executable-ness) of ./buttons-start.sh
+mkdir $SID
+cd $SID
+
+# TODO: conditional on DISTRO: fetch if specified, otherwise ignore
+#   - if no DISTRO to fetch, then this becomes a simple Fusion out of the box, box
+
+gsutil cp gs://buttons-sudosoup/$DISTRO .
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/scripts/start-button.sh]: be sure to change this one as well.

File: projects/buttons/fastener/start-fastener.sh [hunk 9]
```diff
@@ -1,35 +1,37 @@
 #!/bin/bash
 
-NEW_UUID=$(cat /dev/urandom | tr -dc 'a-z0-9' | fold -w 4 | head -n 1)
-ZONE=us-west1-c
+NEW_UUID=$(cat /dev/urandom | env LC_CTYPE=C tr -dc 'a-z0-9' | fold -w 4 | head -n 1)
+ZONE=us-west2-c
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/start-fastener.sh]: is this correct for product. Please confirm/correct.

File: projects/buttons/fastener/start-fastener.sh [hunk 10]
```diff
@@ -39,9 +41,9 @@ gcloud beta compute instances create $NAME-$NEW_UUID \
 --boot-disk-device-name "$NAME-disk-$NEW_UUID" \
 --zone $ZONE \
 --tags http-server,lucid,token-$TOKEN \
---scopes compute-rw \
+--scopes compute-rw,https://www.googleapis.com/auth/cloud-platform \
 --subnet=default $IP --network-tier=PREMIUM \
---service-account labs-209320@appspot.gserviceaccount.com \
+--service-account 215861285408-compute@developer.gserviceaccount.com \
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/start-fastener.sh]: please confirm which is correct.

File: projects/buttons/fastener/start-fastener.sh [hunk 11]
```diff
@@ -59,22 +61,28 @@ pip install --upgrade google-api-python-client
...
+echo "deb https://apt.kubernetes.io/ kubernetes-xenial main" | tee -a /etc/apt/sources.list.d/kubernetes.list
+apt-get update
+apt-get install -y kubectl
+
 #upgrade SSL
 sudo mkdir /usr/local/share/ca-certificates/cacert.org
 sudo wget -P /usr/local/share/ca-certificates/cacert.org http://www.cacert.org/certs/root.crt http://www.cacert.org/certs/class3.crt
 sudo update-ca-certificates
 
+gcloud beta container clusters get-credentials lucidworks-streams-fusion5-cluster --region us-west1 --project labs-3-datastore-dep
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/start-fastener.sh]: Change for correct project and correct cluster in Labs project

File: projects/buttons/fastener/start-fastener.sh [hunk 12]
```diff
@@ -59,22 +61,28 @@ pip install --upgrade google-api-python-client
...
+
 cd /;
-git clone https://github.com/lucidworks/streams
+git clone --single-branch --branch fusion-5-dev https://github.com/sudosoup/streams.git
```
Comment (MarcusSorealheis) [inline projects/buttons/fastener/start-fastener.sh]: point at upstream (Lucidworks' repo)

File: projects/buttons/fusion5_infrastructure/deployment-manager/streams-infra.jinja.schema [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+  version: "v0.1"
+
+imports:
+  - path: "templates/gke_cluster.py"
+  - path: "templates/vpc.py"
+
+properties:
+  region:
+    type: "string"
+    default: "us-west1"
```
Comment (MarcusSorealheis) [inline projects/buttons/fusion5_infrastructure/deployment-manager/streams-infra.jinja.schema]: confirm correct?

File: projects/buttons/fusion5_infrastructure/deployment-manager/templates/gke_cluster.py.schema [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    type: "string"
+    default: "public-subnet"
+    description: "The subnetwork to provision the cluster into"
+  vpc_id:
+    type: "string"
+    default: "default"
+    description: "The network to create the GKE nodes in"
+  region:
+    type: "string"
+    default: "us-west1"
```
Comment (MarcusSorealheis) [inline projects/buttons/fusion5_infrastructure/deployment-manager/templates/gke_cluster.py.schema]: confirm correct.

File: projects/buttons/web/handlers/instancehandlers.py [hunk 15]
```diff
@@ -639,29 +639,55 @@ def post(self, sid=None): # a POST here is a create instance event
...
-                    password = password,
-                    expires = datetime.datetime.now() + datetime.timedelta(0, 604800),
-                    started = datetime.datetime.now()
-                )
-                instance.put()
+                is_alpha = re.compile("^[a-zA-Z]+$")
 
-                slack.slack_message("Instance type %s created for %s!" % (stream.name, user_info.username))
+                # Fusion 5
+                if is_alpha.match(name) and len(name) == 4:
```
Comment (sudosoup) [inline projects/buttons/web/handlers/instancehandlers.py]: need to make sure this matches our final agreed upon length

File: projects/buttons/fastener/main.py (hunk 2 above)
Comment (sudosoup) [reply projects/buttons/fastener/main.py]: this affects an instancehandler.py conditional, see comment there

File: projects/buttons/web/handlers/instancehandlers.py (hunk 15 above)
Comment (MarcusSorealheis) [reply projects/buttons/web/handlers/instancehandlers.py]: this is fine for now

**Pull request merged**